### PR TITLE
Permit additional partner_id formats

### LINKED ISSUE #17403: Support for Commercial Marketplace Azure app tracking IDs
### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

[Customer Usage Attribution](https://aka/ms/customerusageattribution) (CUA) now supports two different types of tracking ID. 

1. Commercial Marketplace Azure apps

    ```text
    pid-<guid>-partnercenter
    ```

    These tracking IDs are created automatically by Partner Center for Azure apps published in the commercial marketplace by partners. The IDs may be copied straight out of Partner Center.

1. Partner generated and registered

    ```text
    pid-<guid>
    ```

    For [other use cases](https://docs.microsoft.com/azure/marketplace/azure-partner-customer-usage-attribution#other-use-cases) of CUA, the partner needs to generate and register a GUID in Partner Center, The GUID is intended to represent the offer and channel.

The Terraform azurerm provider has support for the second option. i.e. `pid-<guid>`. 

* The [partner_id](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#partner_id) argument permits `partner_id = <guid>`
* The value may also be sourced from the `ARM_PARTNER_ID` env var
* partner_id is validated as a GUID
* value is then prefixed with `pid-` and added to the user agent string

The provider does not currently support the `pid-<guid>-partnercenter` format. This is being pushed with Microsoft's ISV partners as the preferred option. 

Suggestion is to expand the validation to permit any of the following:

| partner_id value | user agent string | 
|---|---|
| `<guid>` (current) | `pid-<guid>` |  
| `pid-<guid>` | `pid-<guid>` | 
| `pid-<guid>-partnercenter` | `pid-<guid>-partnercenter` | 

I'm willing to contribute the code myself with testing from @balpurewal and @mormond.

### New or Affected Resource(s)/Data Source(s)

azurerm

### Potential Terraform Configuration

```hcl
provider "azurerm" {
  features {}

  partner_id = "pid-61001a5d-0123-4f37-9d04-314159265359-partnercenter"
}
```

### PR DESCRIPTION
**Pull request by richeney**
Permits the following formats for the partner_id argument in the azurerm provider block:

| partner_id value | user agent string | 
|---|---|
| `<guid>` (current) | `pid-<guid>` |  
| `pid-<guid>` | `pid-<guid>` | 
| `pid-<guid>-partnercenter` | `pid-<guid>-partnercenter` | 

Closes #17403.

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -205,7 +267,7 @@ func azureProvider(supportLegacyTestSuite bool) *schema.Provider {
 			"partner_id": {
 				Type:         schema.TypeString,
 				Optional:     true,
-				ValidateFunc: validation.Any(validation.IsUUID, validation.StringIsEmpty),
+				ValidateFunc: ValidatePartnerID,
```
Comment (katbyte) [inline internal/provider/provider.go:270]: could we revert this change? using .Any makes it clear this can be either a partnerID or emty
```suggestion
				ValidateFunc: validation.Any(ValidatePartnerID, validation.StringIsEmpty),
```

File: internal/provider/provider.go [hunk 2]
```diff
@@ -25,6 +25,68 @@ func TestAzureProvider() *schema.Provider {
 	return azureProvider(true)
 }
 
+func ValidatePartnerID(i interface{}, k string) ([]string, []error) {
+	// ValidatePartnerID checks if partner_id is any of the following:
+	//  * empty string
```
Comment (katbyte) [inline internal/provider/provider.go:30]: can we change this to only validate it is a valid partner id?
```suggestion
```
and not include empty here?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @richeney - just a couple small comments i've left inline

File: internal/provider/provider.go (hunk 2 above)
Comment (richeney) [reply internal/provider/provider.go:30]: Yes, that makes complete sense. I'll commit the suggestion. Thanks @katbyte!

**Review by katbyte — APPROVED** [review]
Thanks @richeney - LGTM 🔑
